polkadot/node/subsystem-util/src/lib.rs

```diff
@@ -439,3 +440,198 @@ impl Validator {
+				maybe_client_features.as_ref(),
+				shuffle,
+				para_id,
+				validator_index,
+			))
+		} else {
+			None
+		}
+	}
+
+	/// Return the per-para chunk index vector if present in the cache.
+	pub fn query_cache_for_para(
+		&mut self,
+		block_number: BlockNumber,
+		session_index: SessionIndex,
+		para_id: ParaId,
+	) -> Option<Vec<ChunkIndex>> {
+		if let Some((shuffle, maybe_client_features)) = self.0.get(&(block_number, session_index)) {
+			let core_start_index = Self::compute_para_start_index(
+				maybe_client_features.as_ref(),
+				shuffle.len(),
+				para_id,
+			);
+
+			let chunk_indices = shuffle
+				.clone()
+				.into_iter()
+				.cycle()
+				.skip(core_start_index)
+				.take(shuffle.len())
+				.collect();
+
+			Some(chunk_indices)
+		} else {
+			None
+		}
+	}
+
+	/// Return and populate the cache with the per-validator chunk index.
+	/// Should only be called if `query_cache_for_validator` returns `None`.
+	pub fn populate_for_validator(
+		&mut self,
+		maybe_client_features: Option<ClientFeatures>,
+		babe_randomness: BabeRandomness,
+		n_validators: usize,
+		block_number: BlockNumber,
+		session_index: SessionIndex,
+		para_id: ParaId,
+		validator_index: ValidatorIndex,
+	) -> ChunkIndex {
+		let shuffle = Self::get_shuffle(
+			maybe_client_features.as_ref(),
+			block_number,
+			babe_randomness,
+			n_validators,
+		);
+		self.0.insert((block_number, session_index), (shuffle, maybe_client_features));
+
+		self.query_cache_for_validator(block_number, session_index, para_id, validator_index)
+			.expect("We just inserted the entry.")
+	}
+
+	/// Return and populate the cache with the per-para chunk index vector.
+	/// Should only be called if `query_cache_for_para` returns `None`.
+	pub fn populate_for_para(
+		&mut self,
+		maybe_client_features: Option<ClientFeatures>,
+		babe_randomness: BabeRandomness,
+		n_validators: usize,
+		block_number: BlockNumber,
+		session_index: SessionIndex,
+		para_id: ParaId,
+	) -> Vec<ChunkIndex> {
+		let shuffle = Self::get_shuffle(
+			maybe_client_features.as_ref(),
+			block_number,
+			babe_randomness,
+			n_validators,
+		);
+		self.0.insert((block_number, session_index), (shuffle, maybe_client_features));
+
+		self.query_cache_for_para(block_number, session_index, para_id)
+			.expect("We just inserted the entry.")
+	}
+
+	fn get_shuffle(
+		maybe_client_features: Option<&ClientFeatures>,
+		block_number: BlockNumber,
+		mut babe_randomness: BabeRandomness,
+		n_validators: usize,
+	) -> Vec<ChunkIndex> {
+		let mut indices: Vec<_> = (0..n_validators)
+			.map(|i| ChunkIndex(u32::try_from(i).expect("validator count should not exceed u32")))
+			.collect();
+
+		if let Some(features) = maybe_client_features {
+			if features.contains(ClientFeatures::AVAILABILITY_CHUNK_SHUFFLING) {
+				let block_number_bytes = block_number.to_be_bytes();
+				for i in 0..32 {
+					babe_randomness[i] ^= block_number_bytes[i % block_number_bytes.len()];
```

Comment:
remind me why we need babe randomness as a seed considering we're shuffling validator indices using it per session already?

Reply:
honestly, I'm no expert on this. I heard it's not a good idea to only use monotonically increasing numbers as seeds in order to get the best quality output from PRNGs.

it would be great if @burdges could have a look at the shuffling algorithm I added and let me know if it's a sane option. Once it's released and enabled via the runtime upgrade, it would be quite difficult to change it (requires a client upgrade on all validators).

In a nutshell here's how it works:
1. Initialise a ChaCha8 rng. The seed is created by xor-ing the babe randomness from the session and the current block number.
2. Shuffle the chunk indices using this rng.
3. Initialise another ChaCha8 rng using the ParaId as the seed and generate one number, let's call it `p`.
4. Perform a rotation of the shuffled chunk indices array using the number from step 3 as the rotation offset, such that the chunk index at position `p` now becomes the first element in the array.

We then use the array from step 4 as the candidate-specific chunk indices array. To find the chunk index for validator `i`, we just get the element with on the `i`-th position.

Reply:
no longer relevant, resolving